### PR TITLE
[Bug] Use integration schemas for required_field types

### LINKED ISSUE #3301: [Bug] Required Fields Type Unknown for some Integration Fields
**Describe the bug**

I recently noticed while trying to support required fields for ESQL rules that we're not correctly populating the field types. When we try to build required_fields build time fields, integration fields that don't overlap ecs/beats/endgame/non-ecs are missing the `type` and report `Unknown`. 

**To Reproduce**
1. Debug view-rule on an integration rule.
2. Breakpoint on the return of `get_required_fields` in **rule.py** 
3. Looked at the required list and see fields with the `Unknown` field type, which should be populated from the integrations schemas.

**Expected behavior**
1. Fields that appear in an integration schema should not be populated with `Unknown`.

**Screenshots**

Screenshot 2023-12-04 at 12 19 48 PM

### PR DESCRIPTION
**Pull request by Mikaayenson**
## Issues

Resolves #3301

## Summary

I made several small refactors to clean up the logic:
1. https://github.com/elastic/detection-rules/commit/b3a9f329be27e376582194957e84928ade96a382 : broke out core logic of `get_integration_schema_data` to be used in other methods
2. https://github.com/elastic/detection-rules/commit/e4b793db63d240dc14e4eca5625b53d9c21e844f : broke out how we parse datasets to be used in other methods
3. https://github.com/elastic/detection-rules/commit/c7382f2e81de5a94349bd9c8365389b401711106 : organized how we obtain datasets and modules into a reusable single method

Finally added support to check integration schemas when building required_fields build time field.

### Testing

Load up an integration rule and verify the build time field types are populated (and not `Unknown`). 

<details><summary>Sample CLI Output</summary>
<p>

```bash

python -m detection_rules view-rule /Users/stryker/workspace/Elastic/detection-rules/rules/integrations/github/defense_
evasion_github_protected_branch_settings_changed.toml --api-format 


Loaded config file: /Users/stryker/workspace/Elastic/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

{
  "author": [
    "Elastic"
  ],
  "description": "This rule detects setting modifications for protected branches of a GitHub repository. Branch protection rules can be used to enforce certain workflows or requirements before a contributor can push changes to a branch in your repository. Changes to these protected branch settings should be investigated and verified as legitimate activity. Unauthorized changes could be used to lower your organization's security posture and leave you exposed for future attacks.",
  "from": "now-9m",
  "index": [
    "logs-github.audit-*"
  ],
  "language": "eql",
  "license": "Elastic License v2",
  "name": "GitHub Protected Branch Settings Changed",
  "query": "configuration where event.dataset == \"github.audit\" \n  and github.category == \"protected_branch\" and event.type == \"change\" \n",
  "related_integrations": [
    {
      "package": "github",
      "version": "^1.0.0"
    }
  ],
  "required_fields": [
    {
      "ecs": true,
      "name": "event.dataset",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "event.type",
      "type": "keyword"
    },
    {
      "ecs": false,
      "name": "github.category",
      "type": "keyword"
    }
  ],
  "risk_score": 47,
  "rule_id": "07639887-da3a-4fbf-9532-8ce748ff8c50",
  "severity": "medium",
  "tags": [
    "Domain: Cloud",
    "Use Case: Threat Detection",
    "Tactic: Defense Evasion",
    "Data Source: Github"
  ],
  "threat": [
    {
      "framework": "MITRE ATT&CK",
      "tactic": {
        "id": "TA0005",
        "name": "Defense Evasion",
        "reference": "https://attack.mitre.org/tactics/TA0005/"
      },
      "technique": [
        {
          "id": "T1562",
          "name": "Impair Defenses",
          "reference": "https://attack.mitre.org/techniques/T1562/",
          "subtechnique": [
            {
              "id": "T1562.001",
              "name": "Disable or Modify Tools",
              "reference": "https://attack.mitre.org/techniques/T1562/001/"
            }
          ]
        }
      ]
    }
  ],
  "timestamp_override": "event.ingested",
  "type": "eql",
  "version": 2
}
```

</p>
</details>  

Note the `github.category` in this example finds the type `keyword` instead of unknown like reported in the issue.

### CODE REVIEW

**Review by terrancedejesus — APPROVED** [review]
Approving.

File: detection_rules/integrations.py [hunk 1]
```diff
@@ -321,38 +319,78 @@ def get_integration_schema_data(data, meta, package_integrations: dict) -> Gener
...
+
+    global NOTIFIED_INTEGRATIONS
+    if notice and data.get("notify", False) and integration not in NOTIFIED_INTEGRATIONS:
+
+        # flag to only warn once per integration for available upgrades
+        NOTIFIED_INTEGRATIONS.add(integration)
+
+        print(f"\n{data.get('name')}")
+        for line in notice:
+            print(line)
```
Comment (brokensound77) [inline detection_rules/integrations.py:360]:
```suggestion
        print('\n'.join(notice))
```

File: detection_rules/integrations.py [hunk 2]
```diff
@@ -321,38 +319,78 @@ def get_integration_schema_data(data, meta, package_integrations: dict) -> Gener
...
                         "stack_version": stack_version, "ecs_version": ecs_version,
                         "package_version": package_version, "endgame_version": endgame_version}
                 yield data
 
 
+def get_integration_schema_fields(integrations_schemas: dict, package: str, integration: str,
+                                  min_stack: Version, packages_manifest: dict,
+                                  ecs_schema: dict, data: dict) -> dict:
+    """Extracts the integration fields to schema based on package integrations."""
+
```
Comment (brokensound77) [inline detection_rules/integrations.py:338]:
```suggestion
```

File: detection_rules/integrations.py [hunk 3]
```diff
@@ -321,38 +319,78 @@ def get_integration_schema_data(data, meta, package_integrations: dict) -> Gener
...
+    package_version, notice = find_latest_compatible_version(package, integration, min_stack, packages_manifest)
+    notify_user_if_update_available(data, notice, integration)
+
+    schema = collect_schema_fields(integrations_schemas, package, package_version, integration)
+    schema.update(ecs_schema)
```
Comment (brokensound77) [inline detection_rules/integrations.py:343]: This means that for any colliding types, ECS schema takes precedence. I remember we discussed this during original implementation. There could also be implications further down this chain (see other comments)

**Review by brokensound77 — COMMENTED** [review]
This looks to include a lot of good cleanup and modularity. Left a few suggestions and comments. 

Also, can you provide a delta of the changes before and after this on just the required fields per rule

- main: make release
- this PR: make release
- for all rules where required fields changed, list

File: detection_rules/integrations.py [hunk 4]
```diff
@@ -321,38 +319,78 @@ def get_integration_schema_data(data, meta, package_integrations: dict) -> Gener
...
+        # cleanup extra quotes pulled from ast field
+        value = value.strip('"')
+
+        integration = 'Unknown'
+        if '.' in value:
+            package, integration = value.split('.', 1)
+        else:
+            package = value
+
+        if package in list(package_manifest):
```
[the remark this answers is not in the document]
Comment (Mikaayenson) [reply detection_rules/integrations.py:388]: If it's not, we typically define those statically in the definitions. That's how we handle things like ml integrations etc.

**Comment by Mikaayenson** [comment]
[8.12_3301.zip](https://github.com/elastic/detection-rules/files/13620221/8.12_3301.zip)
[8.12_main.zip](https://github.com/elastic/detection-rules/files/13620227/8.12_main.zip)
[diff.txt](https://github.com/elastic/detection-rules/files/13620250/diff.txt)


<details><summary>Diff Details</summary>
<p>

```bash

diff --color 8.12/rules/credential_access_google_workspace_drive_encryption_key_accessed_by_anonymous_user.json 8.12_3301/rules/credential_access_google_workspace_drive_encryption_key_accessed_by_anonymous_user.json
48c48
<       "type": "unknown"
---
>       "type": "keyword"
92c92
<   "version": 2
---
>   "version": 3
diff --color 8.12/rules/credential_access_multiple_auth_events_from_single_device_behind_proxy.json 8.12_3301/rules/credential_access_multiple_auth_events_from_single_device_behind_proxy.json
47c47
<       "type": "unknown"
---
>       "type": "keyword"
diff --color 8.12/rules/credential_access_okta_mfa_bombing_via_push_notifications.json 8.12_3301/rules/credential_access_okta_mfa_bombing_via_push_notifications.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "keyword"
77c77
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/credential_access_okta_potentially_successful_okta_bombing_via_push_notifications.json 8.12_3301/rules/credential_access_okta_potentially_successful_okta_bombing_via_push_notifications.json
47c47
<       "type": "unknown"
---
>       "type": "keyword"
52c52
<       "type": "unknown"
---
>       "type": "keyword"
82c82
<   "version": 208
---
>   "version": 209
diff --color 8.12/rules/credential_access_potential_successful_linux_ftp_bruteforce.json 8.12_3301/rules/credential_access_potential_successful_linux_ftp_bruteforce.json
26c26
<       "type": "unknown"
---
>       "type": "keyword"
31c31
<       "type": "unknown"
---
>       "type": "keyword"
104c104
<   "version": 4
---
>   "version": 5
diff --color 8.12/rules/credential_access_potential_successful_linux_rdp_bruteforce.json 8.12_3301/rules/credential_access_potential_successful_linux_rdp_bruteforce.json
26c26
<       "type": "unknown"
---
>       "type": "keyword"
99c99
<   "version": 4
---
>   "version": 5
diff --color 8.12/rules/defense_evasion_github_protected_branch_settings_changed.json 8.12_3301/rules/defense_evasion_github_protected_branch_settings_changed.json
34c34
<       "type": "unknown"
---
>       "type": "keyword"
72c72
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.json 8.12_3301/rules/defense_evasion_google_workspace_new_oauth_login_from_third_party_application.json
49c49
<       "type": "unknown"
---
>       "type": "keyword"
115c115
<   "version": 2
---
>   "version": 3
diff --color 8.12/rules/discovery_denied_service_account_request.json 8.12_3301/rules/discovery_denied_service_account_request.json
36c36
<       "type": "unknown"
---
>       "type": "keyword"
41c41
<       "type": "unknown"
---
>       "type": "keyword"
71c71
<   "version": 4
---
>   "version": 5
diff --color 8.12/rules/discovery_pspy_process_monitoring_detected.json 8.12_3301/rules/discovery_pspy_process_monitoring_detected.json
38c38
<       "type": "unknown"
---
>       "type": "keyword"
99c99
<   "version": 4
---
>   "version": 5
diff --color 8.12/rules/discovery_suspicious_self_subject_review.json 8.12_3301/rules/discovery_suspicious_self_subject_review.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "keyword"
52c52
<       "type": "unknown"
---
>       "type": "keyword"
57c57
<       "type": "unknown"
---
>       "type": "keyword"
87c87
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/execution_shell_via_meterpreter_linux.json 8.12_3301/rules/execution_shell_via_meterpreter_linux.json
31c31
<       "type": "unknown"
---
>       "type": "keyword"
115c115
<   "version": 3
---
>   "version": 4
diff --color 8.12/rules/execution_shell_via_udp_cli_utility_linux.json 8.12_3301/rules/execution_shell_via_udp_cli_utility_linux.json
39c39
<       "type": "unknown"
---
>       "type": "keyword"
133c133
<   "version": 3
---
>   "version": 4
diff --color 8.12/rules/execution_user_exec_to_pod.json 8.12_3301/rules/execution_user_exec_to_pod.json
36c36
<       "type": "unknown"
---
>       "type": "keyword"
41c41
<       "type": "unknown"
---
>       "type": "keyword"
46c46
<       "type": "unknown"
---
>       "type": "keyword"
51c51
<       "type": "unknown"
---
>       "type": "keyword"
81c81
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/initial_access_anonymous_request_authorized.json 8.12_3301/rules/initial_access_anonymous_request_authorized.json
35c35
<       "type": "unknown"
---
>       "type": "keyword"
40c40
<       "type": "unknown"
---
>       "type": "keyword"
45c45
<       "type": "unknown"
---
>       "type": "keyword"
84c84
<   "version": 5
---
>   "version": 6
diff --color 8.12/rules/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.json 8.12_3301/rules/initial_access_google_workspace_object_copied_from_external_drive_and_access_granted_to_custom_application.json
50c50
<       "type": "unknown"
---
>       "type": "keyword"
60c60
<       "type": "unknown"
---
>       "type": "keyword"
103c103
<   "version": 3
---
>   "version": 4
diff --color 8.12/rules/initial_access_multiple_client_addresses_with_single_okta_session.json 8.12_3301/rules/initial_access_multiple_client_addresses_with_single_okta_session.json
53c53
<       "type": "unknown"
---
>       "type": "keyword"
104c104
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/initial_access_sign_in_events_via_third_party_idp.json 8.12_3301/rules/initial_access_sign_in_events_via_third_party_idp.json
43c43
<       "type": "unknown"
---
>       "type": "keyword"
89c89
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/persistence_exposed_service_created_with_type_nodeport.json 8.12_3301/rules/persistence_exposed_service_created_with_type_nodeport.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "keyword"
52c52
<       "type": "unknown"
---
>       "type": "keyword"
83c83
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/persistence_github_org_owner_added.json 8.12_3301/rules/persistence_github_org_owner_added.json
34c34
<       "type": "unknown"
---
>       "type": "keyword"
72c72
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/persistence_kernel_driver_load.json 8.12_3301/rules/persistence_kernel_driver_load.json
27c27
<       "type": "unknown"
---
>       "type": "keyword"
98c98
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/persistence_microsoft_365_teams_custom_app_interaction_allowed.json 8.12_3301/rules/persistence_microsoft_365_teams_custom_app_interaction_allowed.json
62c62
<       "type": "unknown"
---
>       "type": "keyword"
88c88
<   "version": 102
---
>   "version": 103
diff --color 8.12/rules/persistence_organization_owner_role_granted.json 8.12_3301/rules/persistence_organization_owner_role_granted.json
34c34
<       "type": "unknown"
---
>       "type": "keyword"
72c72
<   "version": 1
---
>   "version": 2
diff --color 8.12/rules/privilege_escalation_container_created_with_excessive_linux_capabilities.json 8.12_3301/rules/privilege_escalation_container_created_with_excessive_linux_capabilities.json
38c38
<       "type": "unknown"
---
>       "type": "keyword"
43c43
<       "type": "unknown"
---
>       "type": "keyword"
48c48
<       "type": "unknown"
---
>       "type": "text"
53c53
<       "type": "unknown"
---
>       "type": "keyword"
58c58
<       "type": "unknown"
---
>       "type": "keyword"
104c104
<   "version": 3
---
>   "version": 4
diff --color 8.12/rules/privilege_escalation_pod_created_with_hostipc.json 8.12_3301/rules/privilege_escalation_pod_created_with_hostipc.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "text"
52c52
<       "type": "unknown"
---
>       "type": "boolean"
57c57
<       "type": "unknown"
---
>       "type": "keyword"
103c103
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/privilege_escalation_pod_created_with_hostnetwork.json 8.12_3301/rules/privilege_escalation_pod_created_with_hostnetwork.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "text"
52c52
<       "type": "unknown"
---
>       "type": "boolean"
57c57
<       "type": "unknown"
---
>       "type": "keyword"
103c103
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/privilege_escalation_pod_created_with_hostpid.json 8.12_3301/rules/privilege_escalation_pod_created_with_hostpid.json
37c37
<       "type": "unknown"
---
>       "type": "keyword"
42c42
<       "type": "unknown"
---
>       "type": "keyword"
47c47
<       "type": "unknown"
---
>       "type": "text"
52c52
<       "type": "unknown"
---
>       "type": "boolean"
57c57
<       "type": "unknown"
---
>       "type": "keyword"
103c103
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/privilege_escalation_pod_created_with_sensitive_hostpath_volume.json 8.12_3301/rules/privilege_escalation_pod_created_with_sensitive_hostpath_volume.json
36c36
<       "type": "unknown"
---
>       "type": "keyword"
41c41
<       "type": "unknown"
---
>       "type": "keyword"
46c46
<       "type": "unknown"
---
>       "type": "text"
56c56
<       "type": "unknown"
---
>       "type": "keyword"
102c102
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/privilege_escalation_privileged_pod_created.json 8.12_3301/rules/privilege_escalation_privileged_pod_created.json
36c36
<       "type": "unknown"
---
>       "type": "keyword"
41c41
<       "type": "unknown"
---
>       "type": "keyword"
46c46
<       "type": "unknown"
---
>       "type": "text"
51c51
<       "type": "unknown"
---
>       "type": "boolean"
56c56
<       "type": "unknown"
---
>       "type": "keyword"
102c102
<   "version": 202
---
>   "version": 203
diff --color 8.12/rules/privilege_escalation_suspicious_assignment_of_controller_service_account.json 8.12_3301/rules/privilege_escalation_suspicious_assignment_of_controller_service_account.json
35c35
<       "type": "unknown"
---
>       "type": "keyword"
40c40
<       "type": "unknown"
---
>       "type": "keyword"
45c45
<       "type": "unknown"
---
>       "type": "keyword"
50c50
<       "type": "unknown"
---
>       "type": "keyword"
55c55
<       "type": "unknown"
---
>       "type": "keyword"
93c93
<   "version": 5
---
>   "version": 6

```

</p>
</details>

**Review by brokensound77 — APPROVED** [review]
🚢

**Comment by eric-forte-elastic** [comment]
Code review looks good and additional testing looks good :+1: 

<details><summary>Test Output</summary>
<p>

```shell
detection-rules on  3301-bug-required-fields-type-unknown-for-some-integration-fields [$?] is  v0.1.0 via  v3.10.12 (venv) on  eric.forte 
❯ python -m detection_rules view-rule rules/integrations/github/defense_evasion_github_protected_branch_settings_changed.toml --api-format 

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

{
  "author": [
    "Elastic"
  ],
  "description": "This rule detects setting modifications for protected branches of a GitHub repository. Branch protection rules can be used to enforce certain workflows or requirements before a contributor can push changes to a branch in your repository. Changes to these protected branch settings should be investigated and verified as legitimate activity. Unauthorized changes could be used to lower your organization's security posture and leave you exposed for future attacks.",
  "from": "now-9m",
  "index": [
    "logs-github.audit-*"
  ],
  "language": "eql",
  "license": "Elastic License v2",
  "name": "GitHub Protected Branch Settings Changed",
  "query": "configuration where event.dataset == \"github.audit\" \n  and github.category == \"protected_branch\" and event.type == \"change\" \n",
  "related_integrations": [
    {
      "package": "github",
      "version": "^1.0.0"
    }
  ],
  "required_fields": [
    {
      "ecs": true,
      "name": "event.dataset",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "event.type",
      "type": "keyword"
    },
    {
      "ecs": false,
      "name": "github.category",
      "type": "keyword"
    }
  ],
  "risk_score": 47,
  "rule_id": "07639887-da3a-4fbf-9532-8ce748ff8c50",
  "severity": "medium",
  "tags": [
    "Domain: Cloud",
    "Use Case: Threat Detection",
    "Tactic: Defense Evasion",
    "Data Source: Github"
  ],
  "threat": [
    {
      "framework": "MITRE ATT&CK",
      "tactic": {
        "id": "TA0005",
        "name": "Defense Evasion",
        "reference": "https://attack.mitre.org/tactics/TA0005/"
      },
      "technique": [
        {
          "id": "T1562",
          "name": "Impair Defenses",
          "reference": "https://attack.mitre.org/techniques/T1562/",
          "subtechnique": [
            {
              "id": "T1562.001",
              "name": "Disable or Modify Tools",
              "reference": "https://attack.mitre.org/techniques/T1562/001/"
            }
          ]
        }
      ]
    }
  ],
  "timestamp_override": "event.ingested",
  "type": "eql",
  "version": 2
}

```

</p>
</details>

File: detection_rules/integrations.py [hunk 5]
```diff
@@ -26,6 +26,7 @@
 from .schemas import definitions
 
 MANIFEST_FILE_PATH = Path(get_etc_path('integration-manifests.json.gz'))
+NOTIFIED_INTEGRATIONS = set()
```
[the remark this answers is not in the document]
Comment (brokensound77) [reply detection_rules/integrations.py:29]: could alternatively call it `_notified_integrations`

[the remark this answers is not in the document]
Comment (Mikaayenson) [reply detection_rules/integrations.py:29]: I'll update when I address merge conflicts

**Review by eric-forte-elastic — APPROVED** [review]

**Pull request merged**